### PR TITLE
Adding CountsStorageService to facilitate using the new counts store.

### PR DESCRIPTION
**Pull request by maxsumrall**
As the new counts store is still a work in progress, this is another necessary thing to have as part
 the current effort to start actually using the new counts store.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        return updaterFactory.getUpdater( countsStore, txId );
+    }
+
+    public IndexStatsUpdater indexStatsUpdater()
+    {
+        return indexStatsUpdater;
+    }
+
+    @Override
+    public Updater apply( long txId )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java]: looks like a strange method that just delegate to another public method

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        long[] values = countsStore.get( CountsKeyFactory.indexSampleKey( labelId, propertyKeyId ) );
+        target.write( values[0], values[1] );
+        return target;
+    }
+
+    @Override
+    public void accept( CountsVisitor visitor )
+    {
+        countsStore.forEach( ( countsKey, values ) -> {
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java:98]: traversal is not atomic here, you can see/not see actual values of one atomic state

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        return countsStore.snapshot( txId );
+    }
+
+    @Override
+    public Updater updaterFor( long txId )
+    {
+        return updaterFactory.getUpdater( countsStore, txId );
+    }
+
+    public IndexStatsUpdater indexStatsUpdater()
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java]: This is not used anywhere, so probably all IndexStatsUpdater can be deleted?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java [hunk 4]
```diff
@@ -62,6 +67,53 @@ public InMemoryCountsStore()
...
+        {
+            if ( snapshot != null )
+            {
+                throw new IllegalStateException(
+                        "Cannot alter count store outside of a transaction while a snapshot is processing." );
+            }
+            map.compute( key, ( k, v ) -> {
+                if ( v == null )
+                {
+                    return Arrays.copyOf( delta, delta.length );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java:102]: do we really need to copy always incoming array? that will create tons of array copies, not only here but on all places where we copy and apply updates?
Maybe final data structure to hold values?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java [hunk 5]
```diff
@@ -156,6 +209,12 @@ public CountsSnapshot snapshot( long txId )
         }
     }
 
+    @Override
+    public void forEach( BiConsumer<CountsKey,long[]> action )
+    {
+        map.forEach( action );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java]: atomicity is broken here

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/ReadOnlyCountsStorageService.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.store.counts;
+
+import org.neo4j.kernel.impl.api.CountsVisitor;
+import org.neo4j.kernel.internal.DatabaseHealth;
+import org.neo4j.register.Register;
+
+public class ReadOnlyCountsStorageService implements CountsStorageService
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/ReadOnlyCountsStorageService.java]: not used at all

Comment (maxsumrall) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/ReadOnlyCountsStorageService.java]: Yes, removed for now.

Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/ReadOnlyCountsStorageService.java]: No, I think we should make sure we use it when we start the db readonly.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java (hunk 4 above)
Comment (maxsumrall) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java:102]: I've add this work to the counts store trello card to be addressed later.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/RecordStorageEngine.java [hunk 7]
```diff
@@ -399,6 +404,17 @@ public void init() throws Throwable
         labelScanStore.init();
     }
 
+    public void initFromSnapshot(CountsSnapshot snapshot)
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/RecordStorageEngine.java:405]: i do not really think that RecordStorageEngine is a good place for those 2 methods

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/LogRecoveryInfo.java [hunk 8]
```diff
@@ -17,30 +17,30 @@
...
+import org.neo4j.kernel.impl.store.counts.CountsSnapshot;
 
-import org.neo4j.kernel.impl.store.counts.keys.CountsKey;
-
-public class CountsSnapshot
+public class LogRecoveryInfo
 {
-    final private Map<CountsKey,long[]> map;
-    final private long txId;
+    LogPosition logPosition;
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/LogRecoveryInfo.java]: private fields?

File: community/kernel/src/main/java/org/neo4j/kernel/recovery/PositionToRecoverFrom.java [hunk 9]
```diff
@@ -47,18 +49,20 @@ public PositionToRecoverFrom( LatestCheckPointFinder checkPointFinder )
      * start recovery from
      * @throws IOException if log files cannot be read
      */
-    @Override
-    public LogPosition apply( long currentLogVersion ) throws IOException
+    public LogRecoveryInfo apply( long currentLogVersion ) throws IOException
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/recovery/PositionToRecoverFrom.java:52]: should we just return CheckPoint object here instead? it sounds reasonable that we have a checkpoint to be able to do recovery from them, so why not to just use them? wdyt?

File: community/kernel/src/test/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStoreIntegrationTest.java [hunk 10]
```diff
@@ -178,7 +178,7 @@ public void run()
             for ( int i = 0; i < repeatTimes; i++ )
             {
                 int id = intermediateStateTestManager.getId();
-                long txId = id + ThreadLocalRandom.current().nextLong( 0, 5 );
+                long txId = id; //+ ThreadLocalRandom.current().nextLong( 0, 5 );
```
Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStoreIntegrationTest.java]: hm...:flashlight: looks like should be uncommented?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/DualCountsStoreUpdater.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+{
+    private final CountsTracker.Updater countsTrackerUpdater;
+    private final CountsAccessor.Updater countsStorageUpdater;
+
+    public DualCountsStoreUpdater( long txId, CountsStorageService countsStorageService, CountsTracker countsTracker,
+            TransactionApplicationMode mode )
+    {
+        countsStorageUpdater = countsStorageService.apply( txId );
+        Optional<CountsAccessor.Updater> result = countsTracker.apply( txId );
+        this.countsTrackerUpdater = result.isPresent() ? result.get() : null;
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/DualCountsStoreUpdater.java]: `result.orElse(null)`

File: enterprise/backup/src/main/java/org/neo4j/backup/OnlineBackupKernelExtension.java [hunk 12]
```diff
@@ -70,17 +70,17 @@
     private final GraphDatabaseAPI graphDatabaseAPI;
     private final LogProvider logProvider;
     private final Monitors monitors;
-    private BackupServer server;
+    private volatile BackupServer server;
```
Comment (MishaDemianenko) [inline enterprise/backup/src/main/java/org/neo4j/backup/OnlineBackupKernelExtension.java]: Can you please describe the case when you need this field to be volatile?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/RecordStorageEngine.java (hunk 7 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/RecordStorageEngine.java:405]: Which is a good place for those methods?

Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/RecordStorageEngine.java:405]: lets talk tomorrow about that :)

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/DualCountsStoreUpdater.java (hunk 11 above)
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/DualCountsStoreUpdater.java]: Mix of `Optional` and `null` looks quite strange. Can `CountsStorageService#apply()` return nullable `CountsAccessor.Updater` instead?

Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/DualCountsStoreUpdater.java]: Or another idea: `CountsStorageService#apply()` always returns non-null. When real updater can't be created some empty impl is returned. With this approach no null checks would be needed.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/DatabasePanicEventGenerator.java [hunk 13]
```diff
@@ -36,15 +33,6 @@ public DatabasePanicEventGenerator( KernelEventHandlers kernelEventHandlers )
...
-        executor.execute( new Runnable()
-        {
-            @Override
-            public void run()
-            {
-                kernelEventHandlers.kernelPanic( error, cause );
-            }
-        } );
-        executor.shutdown();
+        kernelEventHandlers.kernelPanic( error, cause ) ;
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/DatabasePanicEventGenerator.java:36]: Why was this change made? Is it relevant to the PR?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsSnapshot.java [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+    public CountsSnapshot( long txId, Map<CountsKey,long[]> map )
+    {
+        this.map = map;
+        this.txId = txId;
+    }
+
+    public CountsSnapshot( long txId )
+    {
+        this( txId, new ConcurrentHashMap<>() );
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsSnapshot.java]: All other maps in this class are not concurrent. Why this one is? Imo if this class should be thread-safe than `Map` should be `ConcurrentMap` everywhere.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+                throw new IllegalStateException( "unexpected counts key " + countsKey );
+            }
+        } );
+    }
+
+    @Override
+    public void initialize( CountsSnapshot snapshot, DatabaseHealth databaseHealth )
+    {
+        countsStore = new InMemoryCountsStore( snapshot, databaseHealth );
+        indexStatsUpdater = new IndexStatsUpdaterFactory().indexStatsUpdater( countsStore );
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java:134]: Can `IndexStatsUpdaterFactory` be removed? Feels like we can create `indexStatsUpdater` directly.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/keys/JustASimpleChannel.java [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+ */
+package org.neo4j.kernel.impl.store.counts.keys;
+
+import java.io.Flushable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.neo4j.kernel.impl.transaction.log.FlushableChannel;
+
+public class JustASimpleChannel implements FlushableChannel
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/keys/JustASimpleChannel.java]: Looks like this class is not used and should be removed.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java [hunk 17]
```diff
@@ -693,7 +706,12 @@ public void rebuildCounts( File storeDir, String versionToMigrateFrom ) throws I
                     countsStoreFiles, true, false, StoreFileType.STORE );
             File neoStore = new File( storeDir, DEFAULT_NAME );
             long lastTxId = MetaDataStore.getRecord( pageCache, neoStore, Position.LAST_TRANSACTION_ID );
-            rebuildCountsFromScratch( storeDir, lastTxId, pageCache );
+            CountsSnapshot snapshot = rebuildCountsFromScratch( storeDir, lastTxId, pageCache );
+            long logVersion = MetaDataStore.getRecord( pageCache, neoStore, Position.LOG_VERSION );
+            long lastCommittedTx = MetaDataStore.getRecord( pageCache, neoStore, Position.LAST_TRANSACTION_ID );
+
+            // write a check point in the log in order to make recovery work in the newer version
+            new StoreMigratorCheckPointer( storeDir, fileSystem ).checkPoint( logVersion, lastCommittedTx, snapshot );
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java]: Probably better to extract method for this line and 688.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointInfo.java [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.transaction.log.checkpoint;
+
+import org.neo4j.kernel.impl.store.counts.CountsSnapshot;
+
+public interface CheckPointInfo
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointInfo.java]: Why this is not a simple class with two fields?

File: community/kernel/src/main/java/org/neo4j/storageengine/api/StorageEngine.java [hunk 19]
```diff
@@ -113,6 +114,14 @@ void createCommands(
      */
     void prepareForRecoveryRequired();
 
+    /**
+     * As part of the simplified counts store. This will initialize the count store in the storage engine
+     * when doing recovery.
+     */
+    void initFromSnapshot( CountsSnapshot snapshot );
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/storageengine/api/StorageEngine.java:121]: Feels strange to have methods related to counts store exposed in storage engine API. Shouldn't it be unaware of things that are underneath it?

Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/storageengine/api/StorageEngine.java:121]: Where to add that instead? Note that a snapshot is pretty general (a long and a map) and unrelated to storage implementation details.

File: enterprise/ha/src/test/java/org/neo4j/ha/TestPullUpdatesApplied.java [hunk 20]
```diff
@@ -152,7 +153,9 @@ public void failed( InstanceId server )
         Thread.sleep( 15000 );
 
         restart( toKill ); // recovery and branching.
-        boolean hasBranchedData = new File( targetDirectory, "branched" ).listFiles().length > 0;
+        File[] branchedFiles = new File( targetDirectory, "branched" ).listFiles();
+        assertNotNull( branchedFiles );
```
Comment (lutovich) [inline enterprise/ha/src/test/java/org/neo4j/ha/TestPullUpdatesApplied.java:157]: `branchedFiles` could be null when `File( targetDirectory, "branched" )` does not exist. Thus `assertNotNull()` looks a bit strange.

Comment (davidegrohmann) [inline enterprise/ha/src/test/java/org/neo4j/ha/TestPullUpdatesApplied.java:157]: If it was `null` and we would have blown up on the `.length` with a NPE. Isn't it better having an assertion `!= null`?

Comment (lutovich) [inline enterprise/ha/src/test/java/org/neo4j/ha/TestPullUpdatesApplied.java:157]: I understand the NPE possibility but `branchedFiles == null` feels like a completely valid case for this test. This is not super important however.

File: community/kernel/src/main/java/org/neo4j/storageengine/api/StorageEngine.java (hunk 19 above)
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/storageengine/api/StorageEngine.java:121]: Unfortunately I can't think of a better place right now.
These are other things that concern me:
- counts store initialization grew through many layers, would be good to keep it more isolated
- we expose not only counts store from `StorageEngine` but also the fact that it is initialized from the outside

Would be good to discuss this thing in person.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java (hunk 2 above)
Comment (maxsumrall) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java:98]: Right. I've changed this to grab the write lock to ensure that the map doesn't change, but we are only using this in areas where I don't expect we are doing updates anyways, e.g. consistency checker.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java (hunk 3 above)
Comment (maxsumrall) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java]: Yes, can be added again if we need it.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java (hunk 5 above)
Comment (maxsumrall) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java]: Same answer to the problem with accept() in CountsStorageService.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/ReadOnlyCountsStorageService.java (hunk 6 above)
Comment (maxsumrall) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/ReadOnlyCountsStorageService.java]: We can put it back when we start using it.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/LogRecoveryInfo.java (hunk 8 above)
Comment (maxsumrall) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/LogRecoveryInfo.java]: Yes.

File: community/kernel/src/main/java/org/neo4j/kernel/recovery/PositionToRecoverFrom.java (hunk 9 above)
Comment (maxsumrall) [inline community/kernel/src/main/java/org/neo4j/kernel/recovery/PositionToRecoverFrom.java:52]: We are also returning the log recovery position which is not related to the checkpoint, so LogRecoveryInfo seems to make more sense at the moment.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStoreIntegrationTest.java (hunk 10 above)
Comment (maxsumrall) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStoreIntegrationTest.java]: 👻 Yes.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/DualCountsStoreUpdater.java (hunk 11 above)
Comment (maxsumrall) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/DualCountsStoreUpdater.java]: It is the updater from the old counts store, CountsTracker, that might return null.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/DatabasePanicEventGenerator.java (hunk 13 above)
Comment (maxsumrall) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/DatabasePanicEventGenerator.java:36]: Was just some funny code.

**Pull request closed without merging**